### PR TITLE
feat: factory

### PR DESCRIPTION
**Pull request by MathisGD**
Oracle factory to ease integrators life and being able to check the code with a call

### CODE REVIEW

File: src/morpho-chainlink/MorphoChainlinkOracleV2Factory.sol [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+contract MorphoChainlinkOracleV2Factory is IMorphoChainlinkOracleV2Factory {
+    /* STORAGE */
+
+    /// @inheritdoc IMorphoChainlinkOracleV2Factory
+    mapping(address => bool) public isChainlinkOracle;
+
+    /* EXTERNAL */
+
+    /// @inheritdoc IMorphoChainlinkOracleV2Factory
+    function createChainlinkOracle(
```
Comment (MerlinEgalite) [inline src/morpho-chainlink/MorphoChainlinkOracleV2Factory.sol:24]:
```suggestion
    function createChainlinkOracleV2(
```

File: src/morpho-chainlink/MorphoChainlinkOracleV2Factory.sol [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            quoteVault,
+            quoteVaultConversionSample,
+            quoteFeed1,
+            quoteFeed2,
+            quoteTokenDecimals
+        );
+
+        isChainlinkOracle[address(oracle)] = true;
+
+        emit CreateChainlinkOracle(msg.sender, address(oracle));
```
Comment (MerlinEgalite) [inline src/morpho-chainlink/MorphoChainlinkOracleV2Factory.sol:52]:
```suggestion
        emit CreateChainlinkOracleV2(msg.sender, address(oracle));
```

File: test/MorphoChainlinkOracleV2FactoryTest.sol [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        bytes32 initCodeHash = hashInitCode(
+            type(MorphoChainlinkOracleV2).creationCode,
+            abi.encode(sDaiVault, 1e18, daiEthFeed, feedZero, 18, vaultZero, 1, usdcEthFeed, feedZero, 6)
+        );
+        address expectedAddress = computeCreate2Address(salt, initCodeHash, address(factory));
+
+        assertFalse(factory.isChainlinkOracle(expectedAddress), "isChainlinkOracle");
+
+        // vm.expectEmit(address(factory));
+        // emit IMorphoChainlinkOracleV2Factory.CreateChainlinkOracle(address(this), expectedAddress);
```
Comment (MerlinEgalite) [inline test/MorphoChainlinkOracleV2FactoryTest.sol:29]: it's still failing?

Comment (MathisGD) [reply test/MorphoChainlinkOracleV2FactoryTest.sol:29]: yes, with a shady compiler error

**Review by Rubilmax — APPROVED** [review]

File: src/morpho-chainlink/interfaces/IMorphoChainlinkOracleV2Factory.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    /// @param quoteVault Quote vault. Pass address zero to omit this parameter.
+    /// @param quoteVaultConversionSample The sample amount of quote vault shares used to convert to underlying.
+    /// Pass 1 if the base asset is not a vault. Should be chosen such that converting `quoteVaultConversionSample` to
+    /// assets has enough precision.
+    /// @param baseFeed1 First base feed. Pass address zero if the price = 1.
+    /// @param baseFeed2 Second base feed. Pass address zero if the price = 1.
+    /// @param quoteFeed1 First quote feed. Pass address zero if the price = 1.
+    /// @param quoteFeed2 Second quote feed. Pass address zero if the price = 1.
+    /// @param baseTokenDecimals Base token decimals.
+    /// @param quoteTokenDecimals Quote token decimals.
```
Comment (QGarchery) [inline src/morpho-chainlink/interfaces/IMorphoChainlinkOracleV2Factory.sol:42]: We can reorder the natspec params too

File: src/morpho-chainlink/interfaces/IMorphoChainlinkOracleV2Factory.sol [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    /// - The vault's sample shares quoted as assets and the base feed prices don't overflow when multiplied.
+    /// - The quote feed prices don't overflow when multiplied.
+    /// - The vault, if set, is ERC4626-compliant.
+    /// @param baseVault Base vault. Pass address zero to omit this parameter.
+    /// @param baseVaultConversionSample The sample amount of base vault shares used to convert to underlying.
+    /// Pass 1 if the base asset is not a vault. Should be chosen such that converting `baseVaultConversionSample` to
+    /// assets has enough precision.
+    /// @param quoteVault Quote vault. Pass address zero to omit this parameter.
+    /// @param quoteVaultConversionSample The sample amount of quote vault shares used to convert to underlying.
+    /// Pass 1 if the base asset is not a vault. Should be chosen such that converting `quoteVaultConversionSample` to
```
Comment (Jean-Grimal) [inline src/morpho-chainlink/interfaces/IMorphoChainlinkOracleV2Factory.sol:35]:
```suggestion
    /// Pass 1 if the quote asset is not a vault. Should be chosen such that converting `quoteVaultConversionSample` to
```

**Review by Jean-Grimal — CHANGES_REQUESTED** [review]

**Review by Jean-Grimal — APPROVED** [review]

File: test/MorphoChainlinkOracleV2FactoryTest.sol (hunk 3 above)
Comment (MathisGD) [reply test/MorphoChainlinkOracleV2FactoryTest.sol]: #78

**Review by QGarchery — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Pull request merged**